src/hotspot/share/prims/jvmtiTagMap.cpp

```diff
@@ -3298,117 +2975,49 @@ void JvmtiTagMap::follow_references(jint heap_filter,
-
-  // if the hashmap is empty then we can skip it
-  if (hashmap->_entry_count == 0) {
-    return;
-  }
-
-  // now iterate through each entry in the table
-
-  JvmtiTagHashmapEntry** table = hashmap->table();
-  int size = hashmap->size();
-
-  JvmtiTagHashmapEntry* delayed_add = NULL;
-
-  for (int pos = 0; pos < size; ++pos) {
-    JvmtiTagHashmapEntry* entry = table[pos];
-    JvmtiTagHashmapEntry* prev = NULL;
-
-    while (entry != NULL) {
-      JvmtiTagHashmapEntry* next = entry->next();
-
-      // has object been GC'ed
-      if (!is_alive->do_object_b(entry->object_raw())) {
-        // grab the tag
-        jlong tag = entry->tag();
-        guarantee(tag != 0, "checking");
-
-        // remove GC'ed entry from hashmap and return the
-        // entry to the free list
-        hashmap->remove(prev, pos, entry);
-        destroy_entry(entry);
-
-        // post the event to the profiler
-        if (post_object_free) {
-          JvmtiExport::post_object_free(env(), tag);
+  bool is_vm_thread = Thread::current()->is_VM_thread();
+  if (!is_vm_thread) {
+    if (num_dead_entries != 0) {
+      JvmtiEnvIterator it;
+      for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+        JvmtiTagMap* tag_map = env->tag_map_acquire();
+        if (tag_map != NULL) {
+          // Lock each hashmap from concurrent posting and cleaning
+          tag_map->unlink_and_post_locked();
         }
-
-        ++freed;
-      } else {
-        f->do_oop(entry->object_addr());
-        oop new_oop = entry->object_raw();
-
-        // if the object has moved then re-hash it and move its
-        // entry to its new location.
-        unsigned int new_pos = JvmtiTagHashmap::hash(new_oop, size);
-        if (new_pos != (unsigned int)pos) {
-          if (prev == NULL) {
-            table[pos] = next;
-          } else {
-            prev->set_next(next);
-          }
-          if (new_pos < (unsigned int)pos) {
-            entry->set_next(table[new_pos]);
-            table[new_pos] = entry;
-          } else {
-            // Delay adding this entry to it's new position as we'd end up
-            // hitting it again during this iteration.
-            entry->set_next(delayed_add);
-            delayed_add = entry;
-          }
-          moved++;
-        } else {
-          // object didn't move
-          prev = entry;
+      }
+      // there's another callback for needs_rehashing
+    }
+  } else {
+    assert(SafepointSynchronize::is_at_safepoint(), "must be");
+    JvmtiEnvIterator it;
+    for (JvmtiEnv* env = it.first(); env != NULL; env = it.next(env)) {
+      JvmtiTagMap* tag_map = env->tag_map_acquire();
+      if (tag_map != NULL && !tag_map->is_empty()) {
+        if (num_dead_entries != 0) {
+          tag_map->hashmap()->unlink_and_post(tag_map->env());
         }
+        // Later GC code will relocate the oops, so defer rehashing until then.
+        tag_map->_needs_rehashing = true;
```

Comment:
Ok, so I'm going to need help to know where in all the different GCs to make this call.

Reply:
For what GC is this wrong? I can see that it might yield more work than required, when performing a full GC, but not that it would do too little work. In other words, I can't see how it is wrong, as opposed to inaccurate. Littering GCs with JVMTI hooks so that we can optimize away an operation we do every young GC, from a full GC, does not really seem worth it IMO.